### PR TITLE
TRUNK-5331: CohortMembership should not require a startDate

### PR DESCRIPTION
**Pull request by patlub**
## JIRA TICKET NAME

[TRUNK-5331: CohortMembership should not require a startDate](https://issues.openmrs.org/browse/TRUNK-5331)
## Issue I worked on
* drop the not-null constraint on CohortMembership.startDate
* remove any Java logic that defaults CohortMembership.startDate to new Date(), and instead default it to null.
* Cohort.intersect should behave in a way that makes sense if you give it {Patient 123, startDate Jan 1} and {Patient 123, startDate Feb 15}.
I would be satisfied if the output is either a cohort with {Patient 123, startDate Feb 15}, or else a cohort with {Patient 123, startDate null}

## Checklist: I completed these to help reviewers :)

- [x] My pull request only contains **ONE single commit**
(the number above, next to the 'Commits' tab is 1).

- [x] My IDE is configured to follow the [**code style**](https://wiki.openmrs.org/display/docs/Java+Conventions) of this project.

- [x] I have **added tests** to cover my changes. (If you refactored
  existing code that was well tested you do not have to add tests)

- [x] I ran `mvn clean package` right before creating this pull request and
  added all formatting changes to my commit.

- [x] All new and existing **tests passed**.

- [x] My pull request is **based on the latest changes** of the master branch.

  No? Unsure? -> execute command `git pull --rebase upstream master`

### CODE REVIEW

File: api/src/main/java/org/openmrs/Cohort.java [hunk 1]
```diff
@@ -267,10 +270,18 @@ public static Cohort union(Cohort a, Cohort b) {
...
-		if (a != null && b != null) {
-			ret.getMemberships().addAll(a.getMemberships());
-			ret.getMemberships().retainAll(b.getMemberships());
+		List<CohortMembership> cohortMemberships = new ArrayList<CohortMembership>();
+			
+		for(CohortMembership  a_member: a.getMemberships()){
+			for(CohortMembership  b_member: b.getMemberships()){
+				if(a_member.getPatientId() == b_member.getPatientId()){
+					if(a_member.getStartDate().before(b_member.getStartDate())){
+						cohortMemberships.add(b_member);
```
Comment (suthagar23) [inline api/src/main/java/org/openmrs/Cohort.java]: Is there any way to check this without two loops? If each one has `x `memberships, then the loop want to  go for `x*x` times to find it.

Comment (patlub) [reply api/src/main/java/org/openmrs/Cohort.java]: Sure, working on this

**Comment by dkayiwa** [comment]
@patlub did you see the travis failure?

File: api/src/main/resources/liquibase-update-to-2.1.xml [hunk 2]
```diff
@@ -110,20 +110,6 @@
 			</column>
 		</addColumn>
 	</changeSet>
-	<changeSet author="vshankar" id="201610242135-3">
-		<preConditions onFail="MARK_RAN">
-			<not>
-				<sqlCheck expectedResult="0">
```
Comment (dkayiwa) [inline api/src/main/resources/liquibase-update-to-2.1.xml]: We do not delete changesets which were released

File: api/src/main/resources/liquibase-update-to-2.2.xml [hunk 3]
```diff
@@ -224,4 +224,10 @@
 								 baseTableName="encounter_diagnosis" baseColumnNames="changed_by"
 								 referencedTableName="users" referencedColumnNames="user_id" />
 	</changeSet>
+	<changeSet id="20180226-TRUNK-5331" author="patrick">
+		<dropNotNullConstraint
```
Comment (dkayiwa) [inline api/src/main/resources/liquibase-update-to-2.2.xml]: All our changesets have preconditions. You can talk to Ola for help

**Comment by djazayeri** [comment]
You need to add more test coverage, including multiple edge cases, for this new intersect method.

File: api/src/main/resources/liquibase-update-to-2.2.xml (hunk 3 above)
Comment (patlub) [reply api/src/main/resources/liquibase-update-to-2.2.xml]: fixed

File: api/src/main/java/org/openmrs/Cohort.java (hunk 1 above)
Comment (patlub) [reply api/src/main/java/org/openmrs/Cohort.java]: @suthagar23 I have tried removing the nested for loops but am hitting a snag. Would you be able to offer up a solution?

**Comment by dkayiwa** [comment]
Closing because @patlub is not working on this any more.
